Dynamic idle and idle related mixer changes

ctzsnooze opened the pull request:
This PR improves idle and throttle behaviour, with particular relevance to dynamic idle and its interactions with throttle scaling and the other throttle related bits of code.

Here is a summary of the changes:

- the DShot idle value that is sent to the motors now stays the same regardless of dynamic idle on/off when linear throttle scaling is active.  Previously the user's idle value would fall when dynamic idle was activated in association with linear throttle scaling.  This has been fixed by re-ordering the throttle adjustments.

- dynamic idle offset is now applied only when needed, instead of being a constant offset, to prevent the deadband that previously occurred at full throttle.  

- the setpoint throttle value sent to the Blackbox records zero throttle when the stick is at zero position, whether dynamic idle is active or not.  The dynamic idle throttle offset is not shown, since it is an internal, technical adjustment required only to make the dynamic idle code work, and doesn't reflect any real change in throttle setpoint.  

- the throttle value sent to the antigravity and dynamic lowpass code now include throttle scaling, but not throttle boost or dynamic idle, to avoid false elevation of the apparent throttle position from dynamic idle that would lead to unnecessary transient changes in thethe AG and lowpass filter cutoffs.

- The dynamic Idle code now uses a modified PID controller during active rpm control phase, with improved performance, and the incoming rpm data is now returned each loop, improving the derivative quality.  In this controller:

-- Smoothed dyn idle D detects the rapid fall in rpm and provides a transient push that is needed in hard chops.  It is filtered heavily and does not add meaningful noise to the motors traces.  Inadequate D may allow a transient drop in rpm immediately after cutting throttle.  Default is 50.
-- The dyn idle P value provides fast control of rpm during the active control phase.  Too much may cause oscillation when dynamic idle is actively controlling the motor rpm.  Not enough may result in rpm drooping below the configured minimum, or slow wobbles from I, just like when P is too low in any PID controller.  In this setting, not enough P is particularly obvious with slow throttle cuts where D isn't strong.  Default is 50.  
-- The dyn idle I value prevents enduring offsets from the set rpm.  It has high 'upwards' gain and falls much more slowly.  Too little and there may be slow rpm wobble during the control phase, or the idle value may rise too slowly; too much may cause a noisy motors trace.  Default is 50.
-- In general, if the control is good but there is noise on the motors trace during the active control phase, the P and I gains can be lowered.  If stronger control is needed and the motors traces are clean, eg with large props, the P and I gains can be increased.

- All the variables used by Dynamic Idle have been re-named with dyn_idle in the name to make it easier to search for them in the code. 

The main benefit of the new dynamic idle control code is that it is more likely than before to prevent desyncs.  For instance, my normal race quad can fly without desync on an idle_min_rpm of only 800rpm (13 rotations a second) and only 1% Dshot idle. 

With this code, when the rpm are not below the minimum, the average motor drive will be set to the DShot idle value.  The `dyn_idle_min_rpm` value is what sets the actual lowest rpm you'll get, not the DShot idle value.  With this code it is quite OK to run the DShot idle value at very low values if you like.

For setups that are not particularly prone to desync, this code lets us use extremely low idle values to enhance hang time, especially inverted hang time, if that is desirable.  The settings mentioned above give great inverted hang time, and the quad bleeds speed off more rapidly in a race setting.  Motors can be slower to start up from lower rpm, so experiment a bit.  The benefit of active control of idle rpm isn't known for sure at this time.

McGiverGim commented on src/main/flight/pid.h:
```diff
@@ -188,11 +188,11 @@ typedef struct pidProfile_s {
-    uint8_t idle_pid_limit;                 // max P
-    uint8_t idle_max_increase;              // max integrated correction
+    uint8_t idle_min_rpm;               // minimum motor speed enforced by the dynamic idle controller
+    uint8_t dyn_idle_p_gain;                // P gain during active control of rpm
```
Different indentation of the comments.

ctzsnooze replied:
Thanks will fix

mikeller commented:
@ctzsnooze:

> removes transient throttle limit / AIRMODE_LPF code, since dynamic idle largely solves that problem, and it has been little used since added some time back, and is off by default

Can you please supply the data to back your claim of 'little used', to allow reviewers to form their own opinion of how much use it has had?

mikeller commented on src/main/cli/settings.c:
```diff
@@ -1152,11 +1150,11 @@ const clivalue_t valueTable[] = {
+    { "dyn_idle_p_gain",            VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = { 1, 250 }, PG_PID_PROFILE, offsetof(pidProfile_t, dyn_idle_p_gain) },
+    { "dyn_idle_i_gain",            VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = { 1, 250 }, PG_PID_PROFILE, offsetof(pidProfile_t, dyn_idle_i_gain) },
+    { "dyn_idle_d_gain",            VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = { 0, 250 }, PG_PID_PROFILE, offsetof(pidProfile_t, dyn_idle_d_gain) },
+    { "dyn_idle_max_increase",      VAR_UINT8 | PROFILE_VALUE, .config.minmaxUnsigned = { 10, 250 }, PG_PID_PROFILE, offsetof(pidProfile_t, dyn_idle_max_increase) },
```
Why was this renamed - this just creates confusion and extra work. Also, why was the upper limit changed?

ctzsnooze replied:
The previous `idle_p` was not in fact a P gain, it was overall integrator gain.  

The previous `idle_adjustment_speed` wasn't a speed, and there was no equivalent for it in the new PID based code.

The defaults for each value of the PIDs are now different.  Having different names in 4.3 avoids problems with cut and paste of inappropriate defaults from 4.2.

In general I felt that the names of all parameters related to dynamic idle should have the form `dyn_idle_xxx` to distinguish them for parameters unrelated to dynamic idle, such as DShot idle.

This makes it much easier to locate all dynamic idle related parameters in the code and for users when they search in CLI.

Other than that `idle_min_rpm` was used in the CMS, and I didn't want to make a change to a CMS value at this time,  I would have renamed it also, since it is only applicable when dynamic idle is active, and could otherwise be confused with DShot idle.  I think it would be best to re-name it also, to `dyn_idle_min_rpm`.

ctzsnooze replied:
The upper limit was changed to permit up to 2.5x greater PID gains.
I'm not entirely sure how much greater the absolute upper limit of PID gain might be.  
The intent is that 100 = 'normal', so that 250 means 2.5x normal, and 25 means 0.25x normal.

ctzsnooze replied:
I've set the defaults back to 50 to allow up to 5x greater values on the upside, and down to off on the downside.
The wide range is for testing and development, most likely will be narrowed once weget more experience with suitability of defaults.

mikeller replied:
Any reason to not allow these values to go to 255 like before?

ctzsnooze replied:
good point, none

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -218,22 +218,38 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+            motorRangeMinIncrease = constrainf((dynIdleP + mixerRuntime.dynIdleI + dynIdleD), 0.0f, maxIncrease);
+            mixerRuntime.motorsIncreased = (motorRangeMinIncrease > 0);
+
+//            DEBUG_SET(DEBUG_DYN_IDLE, 0, (motorRangeMinIncrease * 10000));
```
Please remove commented out code.

ctzsnooze replied:
OK will do

ctzsnooze replied:
I'll leave in the comments that require action before the code is complete, or are necessary for a reviewer to appreciate why something has been done.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -218,22 +218,38 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+            mixerRuntime.motorsIncreased = (motorRangeMinIncrease > 0);
+
+//            DEBUG_SET(DEBUG_DYN_IDLE, 0, (motorRangeMinIncrease * 10000));
+            DEBUG_SET(DEBUG_DYN_IDLE, 0, (MAX(-1000.0f, dynIdleP * 10000)));
```
What is the reason for constraining this debug value? This just means we are wasting logging bandwidth for writing series of '1000' when we could be logging non-meaningless values instead.

ctzsnooze replied:
The P value, while active, is within a range of +/- 1000 of zero, approximately; that is the range we are interested in visualising in the log.  And we want good resolution to see the noise in it.
But the P value goes massively negative when rpm is above minimum, to values like -30,000, or so low as to rotate to a positive value.  Any value below -1000 is of zero interest to us, since the PIDsum will invariably be negative and the motor boost will be zero.
Without this constraint, the view in the blackbox log is of a P trace with very little vertical resolution of the band of interest, because blackbox always sets limits that include all values in the log, and if we have values at -30,000, we see little detail in the 0-500 range.  We can't fix this by changing blackbox scaling percentage either.  I have discussed this with McGyverGim.  Rather than allowing the user to configure % zoom, it would be better for the user to be allowed to set absolute minimum and maximum values for a trace.  Then this would not be necessary.

ctzsnooze replied:
Given that logged values are compressed, there should be less data in the log if we keep logging an identical number rather than one that changes all the time, when we don't care about the actual values.

mikeller replied:
Good point about overruns / underrruns - pretty much all of these debug values should probably be constrained to avoid this (i.e. upper and lower bounds defined with `constrain(x, UINT16_MIN, UINT16_MAX)` or similar).

mikeller replied:
Needs to be constrained to avoid overruns.

mikeller commented on src/main/flight/mixer_init.h:
```diff
@@ -38,11 +38,15 @@ typedef struct mixerRuntime_s {
-    float idleMinMotorRps;
-    float idleP;
-    float oldMinRps;
+    float dynIdleMinRps;
```
Can you please rename these back? There is a lot going on in this pull request already, and having arbitrary renames added into the mix just makes it way harder to see what was actually changed.

ctzsnooze replied:
The new names aren't arbitrary.
They indicate that the parameter is part of the dynamic idle code, and better indicate what they actually do.
I changed them mostly because it makes the code easier to follow.

mikeller replied:
Anyhow, the large-scale rename makes it hard to review this, so I think changing the names back for this pull request is a better approach. If the value of better names outweighs the cost of inconvenience to users when parameters that keep their function / scaling get renamed then this can always be done in a separate pull request later.

ctzsnooze replied:
I apologise for the large-scale renaming.
In most cases, the old names don't match the function of the parameter in the new code, and actually were incorrect in the old code.  
I agree that the two quoted above could be changed back, but is the translation of `idleMaxIncrease` into `dynIdleMaxIncrease `, and  `idleMinMotorRps ` into `dynIdleMinRps` so tough to follow?  
With all new parameters starting with `dynIdle...` or `dyn_idle...`I found the code easier to follow, to be honest.

mikeller replied:
It just makes it hard to figure out what is actual functional change, and what is just a cosmetic rename.

ctzsnooze replied:
the code is functionally completely different, nothing like before, so both have to be sort of evaluated differently.
The old code had a very noisy D signal being added to an error signal, and both integrated; if the integral is positive, then the motors are boosted via a throttle lift.  In a sense the D becomes a P signal after integration, and the error becomes a classic I signal after integration. 
Due to the gating, and noise on D, the outcome wasn't easy to predict.  Spikes on the D element would push the integral briefly upwards, and give an earlier start, after a throttle chop, than one might otherwise get from a simple P I controller.  
The change I made is to use a classical PID controller with classical P, I and D elements.  In practice, D from rpm is very very noisy, and simple addition doesn't work well.  The peak of un-delayed D comes at the steepest part of the rpm fall, and since motors fall sort of exponentially, we get too much D where we don't really want it.  I learned that a simple PI controller alone does an exceptionally good job.  Due to the gating effect of not allowing the output below zero, we can smooth and delay D and add it effectively only at the start, as a kind of kick to pre-empt a fall below setpoint RPM.  In most tests, very little D is actually needed so long as I can be held close to the needed value.  But since 99% of the time, while flying, the RPM is way above this controller's setpoint, a continuously accumulating I would be massively high all the time.  Hence the constraints on I which mostly intend to keep I close to a useful value all the time.  
It wasn't easy for me to appreciate the subtlety of the original code, and very difficult to improve on it.  I found it much easier to understand once it was in familiar P, I and D values.

mikeller commented on src/main/flight/pid.h:
```diff
@@ -188,11 +188,11 @@ typedef struct pidProfile_s {
-    uint8_t idle_p;                         // kP
-    uint8_t idle_pid_limit;                 // max P
-    uint8_t idle_max_increase;              // max integrated correction
+    uint8_t idle_min_rpm;               // minimum motor speed enforced by the dynamic idle controller
```
These should be renamed back to their original names as well I think.

ctzsnooze replied:
`idle_p` wasn't actually what it is named as, it's actually the integral sum kI.
the new code doesn't use a customisable `idle_pid_limit`
naming them consistently for the application they are now used for is appropriate, I think.

ctzsnooze replied:
It's looking likely that a single gain adjustment may be sufficient to modify P and I gain at the same time.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -469,25 +485,49 @@ FAST_CODE_NOINLINE void mixTable(timeUs_t currentTimeUs)
+#ifdef USE_DYN_IDLE
+    // adjust the idle offset out of the logged value
+    if (mixerRuntime.dynIdleMinRps > 0.0f) {
+        mixerThrottle -= dynIdleThrottleOffset;
```
I can see what you are trying to do here, but I think the way you are doing it is wrong - since  throttle scaling, throttle boost, and thrust linear all act in a scaling way on the throttle value, the amount of dynamic idle offset that is part of the throttle value will be different at this point from what was added above, so it cannot simply be subtracted - worst case this will end up with a negative throttle value.
Maybe another way to do this will be to calculate the offset as a fraction of the throttle value when it is added above, and then remove the same fraction here?

ctzsnooze replied:
To be honest I couldn't figure a better way.  I wasn't happy with how I did it, but it worked properly as is.
There is a complex series of throttle manipulations going on here, and the outcome of later manipulations depends on what happened previously.
If there is a tidier way to achieve the same end result, then I'll happily implement it.

ctzsnooze replied:
> @ctzsnooze:
> 
> > removes transient throttle limit / AIRMODE_LPF code, since dynamic idle largely solves that problem, and it has been little used since added some time back, and is off by default
> 
> Can you please supply the data to back your claim of 'little used', to allow reviewers to form their own opinion of how much use it has had?

Not readily, unfortunately.  

It was recommended to not activate it at the same time as rpm filtering, so as rpm filtering becomes more ubiquitous, the scope for it's application becomes less and less.  

Perhaps users who would value retaining it can speak up here or during the beta testing phase.

mikeller replied:
My proposal would be:

```
    float dynIdleThrottleOffset = (1 - throttle) * mixerRuntime.idleThrottleOffset;
    float throttleOffsetRatio = 0;
    if (mixerRuntime.dynIdleMinRps > 0.0f) {
        throttle += dynIdleThrottleOffset;
        throttleOffsetRatio = dynIdleThrottleOffset / throttle;
    }

[do throttle adjustments]

#ifdef USE_DYN_IDLE
    // adjust the idle offset out of the logged value
    if (mixerRuntime.dynIdleMinRps > 0.0f) {
        mixerThrottle *= (1 - throttleOffsetRatio);
    }
```

Still not ideal, as there is an additive component to the throttle adjustments as well.
But at least it avoids underruns.



> It was always a niche bit of code, I know very few people actually using it, and fewer actively benefitting from it.

I find the use of this kind of statement as a rationale for changes to be problematic - it is not known if your sample set is representative of the entirety of Betaflight users, so extrapolating from it for all users is not a sound approach.
If you postulate that this is the case, I think a good approach will be to collect statistics on usage in the first place, and then make a decision on removal of features once we have statistics.

ctzsnooze replied:
Yes I agree, usage statistics would be great.  However we have pressure on flash memory and we could alternately ask around to see how many people actively seem to use it.  I really think it would be OK to take it out at this point and see if people complain about doing that.  Its not hard to put back in if we must.  
The main argument for taking it out is the strong adoption of rpm filtering, and the need to not run both dynamic idle and transient throttle limit at the same time.

mikeller replied:
Usage statistics in Betaflight are collected in configurator, and not in the firmware, so the argument about 'pressure on flash memory' is not applicable.
Again, my proposal is to leave this in, make it accessible in configurator and collect statistics on usage, and then make an informed decision on removal once this has happened.

mikeller reviewed (commented):
I have reviewed what I could - there is a lot going on here. I think it would be easier if you changed the names of parameters / variables back to what they were, as this will remove a considerable amount of change.

ctzsnooze commented:
Some background to dynamic idle in relation to these revisions.

@joelucid developed the original dynamic idle concept - something truly innovative.  The intent was to use DShot telemetry data to improve flight performance at low throttle and minimise the risk of desync.  

This PR provides modest improvements on the original code, with reduced noise transfer into the motor traces (better rpm D filtering), better rpm control accuracy, longer zero motor drive time when possible, easier adjustment, and improved interactions with throttle scaling.  

<b>Without Dynamic Idle</b>

When dynamic idle is inactive, the lowest drive signal sent is the DShot idle percentage.  5% DShot idle roughly corresponds to a 5% PWM signal to the motors, and (in principle, not in practice) 5% of maximum torque.  The motor will generate a constant torque on the prop when motor drive is 'zero', regardless of actual rpm of the motor.  A higher DShot Idle percentage will mean higher idle torque, and higher idle rpm and lift/thrust.  

If the idle torque is insufficient to keep the motor spinning properly at zero throttle, the ESC may desync, causing a sudden loss of control.  Desyncs are most likely to happen during reverse airflow conditions such as:

- during flat drops,
- at start or end of a flip or roll,
- after a sudden 180 degree turn at very high speed

The airflow coming up from below the motor in each case tends to slow the motor down.  If the torque of the airflow exceeds the (fixed) motor torque at idle, the motor may stop rotating or rotate in reverse, causing a desync. Desyncs of this kind are more likely with larger props, over-propped motors, higher ESC PWM frequencies, lower cell count batteries, low DShot idle values, high flip rates and aggressive flying.

Without dynamic idle, the flight controller doesn't know about, and hence cannot respond to, dangerously low rpm.

As a result, the user must configure the DShot idle value sufficiently high to avoid desyncs, but not so high as to make the quad too 'floaty'.  

A big advantage of higher idle is that the motors can accelerate more quickly from a higher initial starting rpm.  Higher idle rpm can therefore improve low throttle authority and responsiveness in flips.  The downside of a high idle will be floaty flat descents, limited inverted hang time, reduced available thrust differential, higher average throttle during flips, and slow deceleration, all because the motors are generating some thrust all the time, even when we actually don't want any at all.

With low idle values, not only is the risk of a desync higher, but motor rpm can drift low, leading to slow startup and impairing performance in flips and stability in flat drops, for example.

There is an 'optimal' idle value where we get reasonable overall behaviour.  That value can vary quite a lot from built to build.  Unfortunately, that value is always a compromise.  For example, we can't have low idle for long hang times without the risk of desync. 

<b>With dynamic idle</b>

With dynamic idle, the rpm of the motors is monitored, using DShot telemetry.  The dynamic idle code tries to stop the rpm of the slowest motor from ever falling below a user configured absolute minimum rpm value.  This provides much better protection of desyncs, even at lower idle values, and reliable spin up despite negative inflow conditions due to stronger idle rpm stability.

Additionally, dynamic idle sets the minimum motor drive percentage to zero.  Under conditions where the PIDs request zero motor drive, the motors can now be sent a zero PWM signal - and generate, effectively, zero thrust - but only if the RPM of the slowest motor is above the configured value.  This can improve braking, flips, inverted hang time, and flat drop fall rates - all situations where positive inflow can keep rpm up even at zero motor drive - without increasing the risk of desync.

It's important to note that under positive inflow conditions, dynamic idle will not let the average motor drive value go to zero.  

With dynamic idle, average motor drive is not allowed below the DShot idle percentage.  

If, for example, DShot idle is set to 5%, and you do an inverted drop, and even if the inflow keeps the motors spinning way above the `idle_min_rpm`, dynamic idle will keep the average motor drive at 5%.  If, during the inverted drop, the PIDs were to request a 10% difference between fastest and slowest motor, the motors can be driven 0% - 10%, and with the guarantee that the rpm will be OK.  Without dynamic idle active, that split would be 5% - 15%, with no guaranteed minimum rpm.  Since the PIDs are always doing something during an inverted drop, we can expect a longer hang time when dynamic idle is active.  We would get an even longer inverted hang time if the DShot idle was set lower again, eg 3%, say, but that's another story (see later).

The dynamic idle code has two functional states.

The first is 'watch and wait', which is the normal state any time the rpm of the slowest motor is above the configured minimum rpm value.  The code does absolutely nothing at this time, other than to try to predict a situation where the motor rpm will, in the near future, fall below the set minimum value (e.g. as throttle is cut).  

To identify a future drop in rpm below the set minimum rpm, the dynamic idle code compares a D signal derived from the rate of fall of rpm to the absolute error, and (hopefully) lifts motor drive just before hitting the minimum.  It must 'look ahead' since motors cannot respond immediately.  Higher `dyn_idle_d_gain` strength causes the motors to kick up earlier, and can be adjusted to higher values if motors have slower response times.  

The second mode is 'full feedback control'.  If the reverse torque from reverse airflow is strong enough that the rpm would be consistently held low, the dynamic idle code provides a PI type control signal that adds motor drive to keep rpm constant, or near constant, at the set minimum value, until the reverse airflow problem goes away, or the pilot throttles up.  In other words, it enters a period of full feedback control of rpm, with the setpoint at the configured minimum rpm.  As soon as the pilot lifts throttle, or if the reverse airflow condition ceases, the controller stops boosting the motors, and quickly reverts to 'watch and wait' mode. 

Like any controller, if the gain is too high, it can oscillate.  The code in this PR allows adjustment of the P and I gain via `dyn_idle_p_gain` and `dyn_idle_i_gain`.  If the rpm value appears to oscillate while in 'full feedback control' mode, these values should be reduced.

<b>Normal configuration</b>

When dynamic idle is active, it's important to understand the interaction between the `idle_min_rpm` value (the lowest allowed dynamic idle rpm value), and the `dshot_idle_value`.

When tested in the motors tab, we can determine the rpm the motors run at for any given DShot percentage.  We may find that:

- the motors start up and spin smoothly at 500 rpm at 1010 in the motors tab, ie 1% DShot idle, and 
- at 5% DShot idle (1050 in the motors tab), they spin at 2500 rpm.  

Let's say we enable dynamic idle at 5%, and set `idle_min_rpm` to 2000 rpm, or about 20% down on the 'nominal' idle rpm.  This is the 'classical' way of configuring dynamic idle.

On arming, the motors will spin at the DShot idle rpm value, and since this is above the `idle_min_rpm` value, the dynamic idle system will be in 'watch and wait mode'.  After takeoff it will stay in 'watch and wait' mode until a motor looks like it might fall below 2000rpm, in which case we now have a 'safety net' that should stop it going much below 2000rpm.  

Additionally, if the PIDs want zero motor drive, the lowest motor can get true zero drive, not the normal 5% Dshot idle drive, providing a greater potential absolute torque range for flips etc and a lower total throttle when the PIDs make small corrections in flat drops, etc.  Preventing excessive falls in rpm leads to more reliable motor spool up from idle under reverse airflow conditions.  

Most of these improvements are, in truth, fairly subtle.  The quad should fly much the same as it would without dynamic idle, with perhaps slightly better flight characteristics, and a reduced chance of a desync.  

<b>Low DShot idle with 'normal' dynamic idle min rpm.</b>

Now let's say we get the same quad, but set the DShot idle value to only 2%, noting that the motors started to spin at 1%, and were at 1000 rpm at 2% drive (1020 in the motors tab).

And let's say we leave the dynamic idle `idle_min_rpm` value at 2000rpm, which is now higher than the DShot idle rpm.

Once airmode is active, the motors will typically go into 'full feedback mode' during flat drops and similar situations, ie when rpm might otherwise fall below 2000rpm.  The dynamic idle code will let the average minimum motor drive fall to 2% when throttle stick is zero, so long as motors at or above 2000 rpm.  This setup will reduce idle thrust, generally, by more than half, providing better inverted drop time, quick flat drops, and quick decelerations.  However, it should not increase the risk of desync, since the minimum idle rpm will be no lower than before. 

This kind of idle configuration is useful for tuning of P and I gains.  If the `DYN_IDLE` debug is used, we can look for P and I oscillations that coincide with oscillations the measured rpm during the 'full feedback control' part of a flat drop.  If present, we should reduce the P and/or I gains.

<b>Low DShot idle values with low dynamic idle min rpm values</b>

Now let's say we get the same quad, with the DShot idle value at 2% (1000rpm), and set the dynamic idle `idle_min_rpm` value to only 800rpm.  This is a seriously low minimum rpm value - much lower than anyone would ordinarily use.  I'm nor recommending you do it, and although my quads can fly quite nicely without desync like this, let's think of this as a theoretical exercise at this point.

This is a fairly extreme configuration.  However, we know from our motors tab testing that the motors will keep spinning provided they don't fall to, or below, 500rpm.  A minimum of 800rpm may still provide us with enough headroom to avoid a desync while flying.  It might be wise to test fly over long grass etc, since we don't really know what the desync threshold is until it actually happens.  We can log to `DYN_IDLE` and check the stability of `minRPS` to see how close we get to 500 rpm during our test flights.

With this setup, the dynamic idle code will still keep the average motor drive percentage at the DShot idle value of 2%.  There will be only a subtle difference over the previous setup during inverted drops or braking because the DSHot idle is still 2%.  However, in flat drops, the quad will fall more quickly because the minimum rpm will be allowed to fall to a lower value.

Because we are letting rpm drop quite low, we may notice a bit of hesitation when throttling up, and we may lose some control authority during some zero throttle inputs.  But sometimes rpm may have dropped as low or even lower without dynamic idle, and since we gain some authority, it may feel much the same.  The only way to know is to give it a try.  

<b>Extremely low DShot idle values and extremely low dynamic idle min rpm values</b>

An even more extreme test of dynamic idle is to now cut DShot idle to a near-zero value.  

If we establish that we don't desync with `idle_min_rpm` set to only 800rpm, we could, in theory, set idle_min_rpm to 800, and set our DShot idle to something like 0.2%.  Een though that DShot idle value can't start the motors spinning, the dynamic idle control should - and does - take over, and keep the motors around 800rpm at idle.  There is a significant risk of desync with such a low minimum rpm value; we are really relying on the performance of the dynamic idle code to survive in this case.  Inverted hang time should be, and is, the longest possible.  

That we can actually fly the quad with effectively no hardware DShot minimum value, using only a minimum idle rpm value, is quite remarkable.  I don't know running setups like this offers any meaningful flight advantage, but it may, especially for freestylers.  This is something we can explore once the newer code is better established.

<b>Tuning and adjustment</b>

The `dyn_idle_d_gain` value should be increased if, during hard throttle cuts, the measured minimum rpm sags briefly to a value that is lower than you'd like.  Going from 100 to 200 doubles the gain, and from 100 to 50 halves the gain.  Increasing gain may be required on very desync prone builds where motors have difficulty accelerating the props at low drive values.  If the rpm drop seems too slow, or the rpm is held up a bit higher than you'd like immediately after a throttle cut, this can be reduced.

`dyn_idle_p_gain` and `dyn_idle_i_gain` can be reduced if there is feedback induced oscillation in rpm during full feedback control.  The rpm is under feedback control whenever the logged rpm iTerm is moving.  When `dyn_idle_i_gain` is a straight line, there is zero motor boost.

`dyn_idle_max_increase` constrains the maximum amount of drive (%) that can be sent to the motors to maintain rpm.  The default of 150 means 15%, which is quite a lot. This value could be increased on very low authority quads where desyncs are a major problem.

ctzsnooze commented:
Here is an example of the dynamic idle PIDs in a throttle chop after a punch upwards followed by a zero throttle flat drop.  Typical kind of 5" race type quad.
DShot idle is at 2% and `idle_min_rpm` is set to 1200 rpm (20 rotations per second).
The debug is DYN_IDLE.
For about 120ms after the throttle chop, one motor is getting zero PWM drive.
The system is in full control mode whenever the cyan `dynamic idle I` trace is active.  It becomes active just before the min rpm gets close to hitting the target rpm.  
The yellow debug line shows actual rotations per second.  While the dynamic idle PI controller is active, rpm doesn't fall much below the intended setpoint of 20 rotations per second. 
The purple line is the derivative factor.  It is this that gets the correction started early.  It is quite smooth due to the heavy lowpass filtering.
![Screen Shot 2020-10-23 at 18 41 00](https://user-images.githubusercontent.com/11737748/96971022-4266a900-1560-11eb-9ce3-8bd825bfbcaf.jpg)

ctzsnooze commented:
Please note that this is very much development code that requires considerable further testing.  My main concern is whether the underlying algorithm is optimal.
Test flights when the dynamic idle rpm is set above a very low DShot idle are not everyone's cup of tea, but they stress the dynamic idle system best, so that's what we need - with a log to the `DYN_IDLE` debug.
The default P, I and D values are now 50, not 100, to allow more gain upwards if required.
Some comments have been addressed.

mikeller commented on src/main/flight/pid.c:
```diff
@@ -198,10 +198,10 @@ void resetPidProfile(pidProfile_t *pidProfile)
-        .idle_p = 50,
-        .idle_pid_limit = 200,
-        .idle_max_increase = 150,
+        .dyn_idle_p_gain = 50,
```
You are changing what individual parameters stands for, and what defaults then need - this means that the parameter group version will have to be incremented, or else unsuspecting users updating the firmware will get invalid settings without ever having touched them.

ctzsnooze replied:
Good point, will do, thanks.

ctzsnooze replied:
PG value incremented to 2

ctzsnooze commented:
@mikeller Now that we are confident that the motors display is accurate, thanks to [472](https://github.com/betaflight/blackbox-log-viewer/pull/472), I am checking how throttle is displayed.

This is with the mixer code in this PR.

DShot idle is 2% and the idle rpm is set so that under full dynamic control the motors get about 5% drive to maintain the set idle rpm.  After a hard throttle cut, the motors should get dynamic idle of 2% until they slow down enough for the dynamic idle rpm limit to be hit, when the feedback system should (and does) maintain rpm.

The image shows on the left with 50% throttle scaling and on the right 100% throttle scaling.

The good:
- the DShot idle value of 2% is respected and not reduced by throttle scaling (in master it would be 1%)
- the dynamic motor control is represented correctly
- the expected motor drive when sticks are full should be half way from 2% to 100%, and is 51%, ie perfect.
- there is no throttle or motor control deadband at the top end of the stick (in master, throttle would max at 1950 RC input in master)

The bad and the ugly:
- throttle is displayed as 47.5% when it should be 50% at 2000 of RCcommand at 50% throttle scaling.
- motor drive isn't quite 100% on full throttle with 100% (no) throttle scaling.

![Screen Shot 2020-10-31 at 08 55 49](https://user-images.githubusercontent.com/11737748/97761598-37aca500-1b5a-11eb-963f-3e7fd533fd23.jpg)

How we display `throttle setpoint` depends somewhat on what the user would expect to see.  

Historically we have not displayed the throttle changes and arising from airmode in throttle setpoint.  I think that is correct and appropriate.

We have to consider whether we show the throttle manipulations around idle that arise from the dynamic idle code.  They are a kind of internal technical nature, very similar to airmode in that respect.

In the image above, from the code of this PR, dynamic idle is not shown elevating throttle, which I think is consistent with the user expectation, since their stick position is hard zero.  On that basis I feel that the throttle offset that provides the dynamic idle boost should not be included in throttle setpoint, much as we don't include the airmode throttle offset.  

The alternative proposition would be that throttle would be shown at 0% until it was lifted to approximately 5% by dynamic idle (in this example); it  would show the jitter and variability arising from the dynamic idle correction.  This could be useful for tuning purposes, but my view is that it will be confusing to ordinary users of the log.  For tuning the debug is more useful since it shows the contribution of each parameter to that jitter.  

Of the two options, I think we should treat the dynamic idle offset in the same way we do airmode, and not display it in throttle setpoint, and that's how this PR is constructed.

In contrast, throttle scaling and limiting really do affect how we intend our throttle to drive the motors.  They are a bit like applying an RC curve to throttle. I think that most users would therefore expect to see the effect of throttle scaling, once a user makes changes, in their throttle setpoint. 

Throttle boost is also useful when included in throttle setpoint, so that when a user configures it, they can see its action, and 'tune' the amount they want..  I think we should retain this.

I'd be grateful for your views on the goals here.  I'll try to correct the mid-throttle offset error in displayed setpoint, not really sure how that arises.

ctzsnooze commented:
I think I've resolved all the interactions with throttle; used a `MAX` sot the offset for dynamic idle is only active when needed.

There are some inevitable issues related to thrust linear but have confirmed the maths is correct.  It slightly elevates the idle drive when set high but not a large amount (and can't easily be avoided).  Its throttle compensations are correct and checked.

With dynamic idle on, settings as above, throttle and motors now track each other perfectly (with TL off, and no throttle boost):

![Screen Shot 2020-10-31 at 14 36 19](https://user-images.githubusercontent.com/11737748/97770386-e8c93480-1b86-11eb-9e2a-2f7a142dda28.jpg)

Here the user can see the impact of throttle boost, showing the extra rise in throttle early, and how, on chopping throttle, the output hits zero ahead of RC command.  Dynamic idle behaviour is unaffected.

![Screen Shot 2020-10-31 at 14 40 53](https://user-images.githubusercontent.com/11737748/97770435-5ecd9b80-1b87-11eb-8729-a4e5e2760bc8.jpg)

This shows Thrust Linear of 50 over the top of dynamic idle.  There are some interactions here worth noting.  Thrust linear of 50 is a lot; without a compensatory throttle reduction, motor drive would increase to 56.25%.  Showing the throttle compensation factor in throttle setpoint would confuse matters considerably, so I don't.  This image shows a throttle push scaled 50% and the resulting small additional boost in motor drive after throttle compensation.  Note that there is also a small residual idle boost, but not much, and I think it is inevitable that the two interact when dynamic idle is active, since the compensation varies depending on throttle value.

![Scale50_TL50_Dshot](https://user-images.githubusercontent.com/11737748/97770521-211d4280-1b88-11eb-96eb-736f5321620b.jpg)

ctzsnooze commented:
Updated and rebased, PG value incremented.
Still have queries (indicated in comments):
- should yaw spin recovery, launch control and GPS return to home act before, or after, airmode?  I think yes for GPS return to home, but not so sure about the other two.
- generally, is the yaw spin recovery stuff in the optimal places
- where should the block of code `// Find roll/pitch/yaw desired output` optimally go?  Perhaps just before the airmode code?  seems a bit 'lost' where it is.
Otherwise, apart from perhaps optimising the dynamic idle default gains, from my perspective this code should be good to go.
I would value flight testing and a really close check of the code changes.

ctzsnooze commented:
Made a significant improvement in the dynamic idle code, wherein the I value falls much more slowly than it rises.  This greatly improves the amount of I that can be applied, so defaults have changed also.  The slow rate of fall of I means that during drops with movement the I value stays close to where it is needed, rather than dropping to zero and delaying the motor response to low rpm.  Default PIDs modified accordingly.
This is a 5" race quad with DShot idle set to only 2%, dynamic idle set to 2000 RPM (equivalent of about 4% DShot idle on this quad).  Hence the dynamic idle
 is above DShot idle, and the motors are pretty much in full dynamic idle control whenever throttle is low.
The image is 10s flight with a flip and multiple throttle pumps and drops. 
The minRPS debug in orange shows rpm of the lowest motor.  Should be 333rps (= 2000rpm).  Remarkable idle control with minimal jitter or noise added to the motors.
![Screen Shot 2020-11-03 at 20 14 39](https://user-images.githubusercontent.com/11737748/97967880-f514f200-1e11-11eb-9ccf-75bebb674d0e.jpg)

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -216,22 +216,28 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+        if (mixerRuntime.dynIdleMinRps > 0.0f) {
+            const float maxIncrease = isAirmodeActivated() ? mixerRuntime.dynIdleMaxIncrease : 0.04f;
+            float minRps = rpmMinMotorFrequency();
+            DEBUG_SET(DEBUG_DYN_IDLE, 3, (minRps * 10));
```
Needs to be constrained to avoid overruns.

ctzsnooze replied:
see below.  60000 rpm is about as fast as a motor goes, that's 1000 fps.  A motor would have to spin 180,000 rpm to get close to an overrun in the debug.  Hence likely doesn't require a constraint.

mikeller commented on src/main/cli/settings.c:
```diff
@@ -1142,9 +1142,7 @@ const clivalue_t valueTable[] = {
 #endif
-#ifdef USE_AIRMODE_LPF
-    { "transient_throttle_limit",   VAR_UINT8 | MASTER_VALUE, .config.minmax = { 0, 30 }, PG_PID_PROFILE, offsetof(pidProfile_t, transient_throttle_limit) },
-#endif
```
This has got nothing to do with dynamic idle and does not belong into this pull request, see https://github.com/betaflight/betaflight/blob/master/.github/pull_request_template.md:

> Keep your pull requests as small and concise as possible. One pull request should only ever add / update one feature. If the change that you are proposing has a wider scope, consider splitting it over multiple pull requests. In particular, pull requests that combine changes to features and one or more new targets are not acceptable.

ctzsnooze replied:
The removal of transient throttle limit does have relevance to dynamic idle, for two reasons:
- transient throttle limit was proposed originally to help deal with an issue of noise on motor traces around zero throttle.  Dynamic idle addresses this problem in a different way, and provides other advantages, so people currently using transient throttle limit should switch to dynamic idle. 
- activating transient throttle limit is not recommended while dynamic idle is active, since both influence motors in an upwards direction and will be additive (unnecessarily so).
Hence it would likely result in fewer user configuration problems if we just removed transient throttle limit.
Additionally transient throttle limit was more of a 'test' feature, for evaluation, and was not proposed with certainty that it would provide any significant improvement.  It is off by default, and we've not had any positive feedback indeed no feedback largely because most people don't use it and if they do they note no benefit.  Thorsten Laux originally wrote it and he no longer uses it or recommends others using it. 
Finally I see this as an attempt to reduce 'bloatware' with features that get little or no use, or have very limited benefit.  
Now if after consideration of what I've written above, you still really do want me to take these changes out and put them in a separate PR, just confirm and I'll do that.

ctzsnooze replied:
The other reason is that because it potentially interacts with other idle related and throttle related functions, it's quite a lot of work to be sure that transient throttle limit works as it should after all these other changes.  It will have to be tested when each of the other throttle modifiers are active.  Ideally it will need to be actively excluded if dynamic idle is active - I won't know how to do this.  All that is a lot of work for what I think is no useful gain.  I honestly don't want to have to go down that path while being of the view that it is a waste of my time.  If we take it out, and people really want it back, then OK, we can put it back and ensure its in the right place and do the testing to ensure it doesn't have adverse interactions.  Otherwise please lets just take it out.

ctzsnooze replied:
Restored transient throttle limit.  It's at the end of the chain as an air mode type throttle lift modifier.
Maybe in a separate PR we can lock it out so it will be inactive when dynamic idle is active.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -216,22 +216,28 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+            motorRangeMinIncrease = constrainf((dynIdleP + mixerRuntime.dynIdleI + dynIdleD), 0.0f, maxIncrease);
+
+            DEBUG_SET(DEBUG_DYN_IDLE, 0, (MAX(-1000.0f, dynIdleP * 10000)));
+            DEBUG_SET(DEBUG_DYN_IDLE, 1, (mixerRuntime.dynIdleI * 10000));
```
Needs to be constrained to avoid overruns.

mikeller commented on src/main/flight/mixer.c:
```diff
@@ -216,22 +216,28 @@ static void calculateThrottleAndCurrentMotorEndpoints(timeUs_t currentTimeUs)
+
+            DEBUG_SET(DEBUG_DYN_IDLE, 0, (MAX(-1000.0f, dynIdleP * 10000)));
+            DEBUG_SET(DEBUG_DYN_IDLE, 1, (mixerRuntime.dynIdleI * 10000));
+            DEBUG_SET(DEBUG_DYN_IDLE, 2, (dynIdleD * 10000));
```
Needs to be constrained to avoid overruns.

ctzsnooze replied:
I'm not sure how to do that. 
Some debugs have the form `lrintf(DEBUG...)` - I'm never sure when/if this is required or it would provide constraint in some way.
At present the values in logs aren't close to rolling over.  P I haven't seen over 1000, don't think it can get close to 32000 even at max gain since there is a limit to how large the error can be (rps can't go below zero).  D can't change very fast, because motors only change slowly, typical values are under 500 max, at full gain on D the number should be well within the limit. I is similar to P.  minRPS also is of limited range. 
I'm not sure its absolutely necessary to constrain them to integer limits, but if you feel strongly, and show me how, I will.

mikeller commented on src/main/flight/mixer_init.c:
```diff
@@ -415,7 +418,9 @@ void mixerInit(mixerMode_e mixerMode)
 
 #ifdef USE_DYN_IDLE
-    mixerRuntime.idleThrottleOffset = getDynamicIdleOffset();
+    mixerRuntime.idleThrottleOffset = motorConfig()->digitalIdleOffsetValue * 0.0001f;
```
Why was this changed to not use the getter?

ctzsnooze replied:
The getter was new, I think you added it, and I must have stuffed up the rebasing.  Sorry.  Will fix.

ctzsnooze replied:
@mikeller - fixed now

mikeller commented on src/main/flight/rpm_filter.c:
```diff
@@ -173,6 +173,7 @@ FAST_CODE_NOINLINE void rpmFilterUpdate()
         if (motor < 4) {
             DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequency[motor]);
         }
+        motorFrequency[motor] = erpmToHz * filteredMotorErpm[motor];
```
What is the rationale for this change?

ctzsnooze replied:
The old way resulted in intermittent rpm values, only when there was a significant change.  This led to a great deal of jitter on the derivative of the RPS signal.  I discussed it with Thorsten and he recommended this change, which definitely improves matters.

ctzsnooze commented:
This PR should be good for testing; here are binaries that can be used, based on build 8050ecd1e.

When testing, please use debug mode `DYN_IDLE` to review the P, I and D values, and the rpm of the slowest motor (which is shown in blackbox as RPS; to convert to RPM, multiply by 60).

Comparisons without dynamic idle should use the debug mode `DSHOT_RPM_TELEMETRY`.

The latest nightly betaflight blackbox log viewer is needed to properly visualise the DShot motor commands.

Remember: enable dynamic idle with `set dyn_idle_min_rpm = NN` in the CLI.  

If your DShot idle value is set very low, ie just above barely spinning (eg, less than 1%), then the dynamic idle minimum rpm becomes the primary idle control value.  If set too low you may get partial desyncs at the end of fast flips or rolls, seen as a kind of wobble or partial loss of control.  In most cases the dynamic idle can be very low, as low as 1000 - 1500 rpm.  However, some ESCs will stall a motor much more readily than others, and will need a higher minimum rpm value.

For example, `set dyn_idle_min_rpm = 15`, with DShot idle of 1%, will:
- enable dynamic idle with a minimum of 1000 rpm, 
- allow zero DShot motor drive if requested by the mixer, unless rpm falls below 1000, and
- not permit an average motor drive value less than 1%, regardless of minimum rpm

Here are test binaries:

[betaflight_4.3.0_STM32F745_8050ecd1e.zip](https://github.com/betaflight/betaflight/files/5587089/betaflight_4.3.0_STM32F745_8050ecd1e.zip)
[betaflight_4.3.0_STM32F7X2_8050ecd1e.zip](https://github.com/betaflight/betaflight/files/5587092/betaflight_4.3.0_STM32F7X2_8050ecd1e.zip)
[betaflight_4.3.0_STM32F411_8050ecd1e.zip](https://github.com/betaflight/betaflight/files/5587093/betaflight_4.3.0_STM32F411_8050ecd1e.zip)
[betaflight_4.3.0_STM32F405_8050ecd1e.zip](https://github.com/betaflight/betaflight/files/5587095/betaflight_4.3.0_STM32F405_8050ecd1e.zip)
